lib/ansible/modules/cloud/vmware/vmware_cfg_backup.py

```diff
@@ -0,0 +1,217 @@
+        choices: ['saved', 'absent', 'loaded']
+extends_documentation_fragment: vmware.documentation
+'''
+
+EXAMPLES = '''
+#save the ESXi configuration locally
+- name: ESXI backup test
+  local_action:
+      module: vmware_cfg_backup
+      hostname: esxi_host
+      username: user
+      password: pass
+      state: saved
+      dest: /tmp/
+'''
+
+RETURN = '''
+dest_file:
+    description: The full path of where the file holding the ESXi configurations was stored
+    returned: changed
+    type: string
+    sample: /tmp/configBundle-esxi.host.domain.tgz
+'''
+
+import os
+
+try:
+    from pyVmomi import vim, vmodl
+    HAS_PYVMOMI = True
+except ImportError:
+    HAS_PYVMOMI = False
+
+
+class VMwareConfigurationBackup(object):
+    def __init__(self, module):
+        self.module = module
+        self.state = self.module.params['state']
+        self.dest = self.module.params['dest']
+        self.src = self.module.params['src']
+        self.hostname = self.module.params['hostname']
+        self.username = self.module.params['username']
+        self.password = self.module.params['password']
+        self.validate_certs = self.module.params['validate_certs']
+        self.content = connect_to_api(self.module)
+        self.host = self.find_host_system()
+
+    def find_host_system(self):
+        host_system = get_all_objs(self.content, [vim.HostSystem])
+        return host_system.keys()[0]
+
+    def process_state(self):
+        if self.state == 'saved':
+            self.save_configuration()
+
+        if self.state == 'absent':
+            self.reset_configuration()
+
+        if self.state == 'loaded':
+            self.load_configuration()
+
+    def load_configuration(self):
+        if not os.path.isfile(self.src):
+            self.module.fail_json(msg="Source file {} does not exist".format(self.src))
+
+        url = self.host.configManager.firmwareSystem.QueryFirmwareConfigUploadURL()
+        url = url.replace('*', self.hostname)
+        #find manually the url if there is a redirect because urllib2 -per RFC- doesn't do automatic redirects for PUT requests
+        try:
+            request = open_url(url=url, method='HEAD', validate_certs=self.validate_certs)
+        except HTTPError as e:
+            url = e.geturl()
+
+        try:
+            with open(self.src, 'rb') as file:
+                data = file.read()
+            request = open_url(url=url, data=data, method='PUT', validate_certs=self.validate_certs,
+                               url_username=self.username, url_password=self.password, force_basic_auth=True)
+        except Exception as e:
+            self.module.fail_json(msg=str(e))
+
+        if not self.host.runtime.inMaintenanceMode:
+            self.enter_maintenance()
+        try:
+            self.host.configManager.firmwareSystem.RestoreFirmwareConfiguration(force=True)
+            self.module.exit_json(changed=True)
+        except Exception as e:
```

Comment:
Also, add ```vmodl.fault.SystemError``` here,

Reply:
Hi @Akasurde . Do you mean than only the vmodl.fault.SystemError needs to exit maintenance?

Reply:
No, I mean

```
except (vmodl.fault.SystemError, Exception) as e:
```

Reply:
Thanks for the clarification Akasurde. However, I am pretty sure that "Exception" catches the "vmodl.fault.SystemError" too.

Reply:
ok.